engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/SearchPredicateFactory.java

```diff
@@ -348,7 +368,22 @@ PredicateFinalStep withParameters(
 	 * @return A new predicate factory using the given object field as root.
 	 */
 	@Incubating
-	SearchPredicateFactory withRoot(String objectFieldPath);
+	SearchPredicateFactory<SR> withRoot(String objectFieldPath);
+
+	/**
+	 * Create a new predicate factory whose root for all paths passed to the DSL
+	 * will be the given object field.
+	 * <p>
+	 * This is used to call reusable methods that apply the same predicate
+	 * on different object fields that have same structure (same sub-fields).
+	 *
+	 * @param objectFieldReference The reference representing the path from the current root to an object field that will become the new root.
+	 * @return A new predicate factory using the given object field as root.
+	 */
+	@Incubating
+	default SearchPredicateFactory<SR> withRoot(ObjectFieldReference<? super SR> objectFieldReference) {
+		return withRoot( objectFieldReference.absolutePath() );
+	}
```

Comment:
Even though I've added these `withRoot(..)` methods to factories, I am not sure that they make much sense with the approach references are taking...

Assuming we somehow address the paths, the factory would still point to the same `<SR>` root type, as the properties in the embedded objects of the "generated" metamodel will use them:

```java
public static class EntityA_ {
	public static ValueFieldReference1<EntityA_, String, String, String> stringA;

	public static SomeNestedObject_ someNestedObject;

	public static RootReferenceScope<EntityA_, EntityA> scope;

	static {
		stringA = ValueFieldReference1.of( "stringA", EntityA_.class, String.class, String.class, String.class );
		someNestedObject = new SomeNestedObject_();

		scope = RootReferenceScopeImpl.of( EntityA_.class, EntityA.class );
	}

	private static class SomeNestedObject_ implements ObjectFieldReference<EntityA_> {

		public final ValueFieldReference1<EntityA_, String, String, String> nestedStringA;

		public SomeNestedObject_() {
			this.nestedStringA = ValueFieldReference1.of( "someNestedObject.stringA", EntityA_.class, String.class, String.class, String.class );
		}

		@Override
		public String absolutePath() {
			return "someNestedObject";
		}

		@Override
		public Class<EntityA_> scopeRootType() {
			return EntityA_.class;
		}
	}
}
```

If we go with the different `<SR>` for `nestedStringA` then we wouldn't be able to use it in a regular case e.g.:
```java
...
.where( f -> f.match().field( EntityA_.someNestedObject.nestedStringA ).matching( "a" ) )
```
Also, if we change the `SR` on `withRoot` call, the current idea for these embedded objects to create a class-per-field would not fly 😔. 

If we ignore the `<SR>` type change, then any property can be passed to the factory methods, which makes the static metamodel pointless 😔 (in this case)

Reply:
Ideally, I'd expect after using `withRoot()`, people using the factory would use references such as `SomeNestedObject_.nestedStringA`, with `nestedStringA` being a *static* field. Just like they would on root types.

So you'd have two classes per object field:

* `SomeNestedObject$Absolute` (the name doesn't matter much, but must not conflict with other classes), which would be the type of `EntityA_.someNestedObject`, and which would expose non-static "sub" fields of type `ValueFieldReference<EntityA_, ...>`.
* `SomeNestedObject_` which would be the type users of `withRoot()` would rely on, and which would expose non-static "sub" fields of type `ValueFieldReference<SomeNestedObject_, ...>`.

I don't think you need an inheritance relationship between these two classes, but you may need to redefine `ObjectFieldReference` to have two generic type parameters: the first being its "holding" type, and the second being the `SR` of factories obtained by calling `withRoot`.

That being said... Since `SomeNestedObject_` is completely specific to the `nestedString` field of `EntityA`, I'm wondering how useful `withRoot()` would be. The main point of `withRoot` is to obtain a factory that can be passed to some util method, which is then able to build a predicate. That method is supposed to be used with different object fields that are assumed to have a similar schema.
If that method works for a specific object field only (with a specific metamodel type, e.g. `SomeNestedObject_`), then... what's the point?

I suppose it all comes down to the "mapped superclass"/inheritance support I mentioned the other time. If we could make sure that object field metamodel uses inheritance, i.e. the metamodel type `EntityB_.SomeNestedObject_` extends `EntityA_.SomeNestedObject_` as soon as `EntityB` extends `EntityA`, then you'd at least be able to create a root from `EntityB_.someNestedObject` and then use `EntityA_.SomeNestedObject_.nestedStringA` on the resulting factory.

Union types at the object field level could be a thing too, but that seems... hard. We'd essentially need `SomeNestedObject_` to extend all its union types. That means multiple inheritance, which means `SomeNestedObject_` must be an interface. Possible, since interfaces can have static fields, but needs investigation.

Reply:
Hmm ok I've come up with the next example. Assuming we have a model as
```java
@Indexed
	public static class EntityA {

		@IndexedEmbedded(includePaths = { "a", "b" })
		public EmbeddedThing embA;

		@IndexedEmbedded(includePaths = { "b", "c" })
		public EmbeddedThing embB;
	}

	public static class EmbeddedThing {
		@FullTextField
		String a;
		@FullTextField
		String b;
		@FullTextField
		String c;
	}
```
and want to have an util to filter by `smth.b` using this `withRoot` approach...

we can generate the metamodel as:
```java
	public static class EntityA_ {

		public static a_ embA;
		public static b_ embB;

		// These two are for the absolute paths: 
		public static class a_ extends ObjectFieldReference<EntityA_, EntityA_a_>{
			public static ValueFieldReference<EntityA_, String, String, String> a; // path = embA.a
			public static ValueFieldReference<EntityA_, String, String, String> b; // path = embA.b
		}

		public static class b_ extends ObjectFieldReference<EntityA_, EntityA_b_>{
			public static ValueFieldReference<EntityA_, String, String, String> b; // path = embB.b
			public static ValueFieldReference<EntityA_, String, String, String> c; // path = embB.c
		}
	}

	// These two are for "withRoot"/  relative paths 
	public static class EntityA_a_ {
		public static ValueFieldReference<EntityA_a_, String, String, String> a; // path = a
		public static ValueFieldReference<EntityA_a_, String, String, String> b; // path = b
	}

	public static class EntityA_b_ {
		public static ValueFieldReference<EntityA_b_, String, String, String> b; // path = b
		public static ValueFieldReference<EntityA_b_, String, String, String> c; // path = c
	}
```
That won't help much for the with `withRoot` ... so we can try:
```java
	// ideally I guess we'd want to generate:
	public static class EntityA_a_b_common_ {
		public static ValueFieldReference<EntityA_a_b_common_, String, String, String> b; // path = b
	}

	public static class EntityA_a_ extends EntityA_a_b_common_ {
		public static ValueFieldReference<EntityA_a_, String, String, String> b; // path = b
	}
	
	public static class EntityA_b_ extends EntityA_a_b_common_ {
		public static ValueFieldReference<EntityA_b_, String, String, String> c; // path = c
	}
	
	// and then the util method could use  EntityA_a_b_common_ with access to `b` 
	//   and we pass withRoot(EntityA_.embA) withRoot(EntityA_.embB) to the util method ... 
```

that could probably work.. but if I understand it correctly... currently that withRoot is not tied to a particular embedded type, meaning we can have:
```java
public static class EntityA {
	@IndexedEmbedded
	EmbeddedThing1 e1;
	@IndexedEmbedded
	EmbeddedThing1 e2;
}

public static class EmbeddedThing1 {
	@FullTextField
	String a;
	// some other fields
}

public static class EmbeddedThing2 {
	@FullTextField
	String a;
	// some other fields maybe different from EmbeddedThing1
}

public static class EntityB {
	@IndexedEmbedded
	EmbeddedThing3 e3;
}

public static class EmbeddedThing3 {
	@FullTextField
	String a;
	// some other fields maybe different from EmbeddedThing1/EmbeddedThing2
}

//util method
private SearchPredicate matchFirstAndLastName(SearchPredicateFactory f, String avalue) {
	return f.match().field( "a" ).matching( avalue ).toPredicate();
}
```
it is ok now to pass any of 
```java
withRoot("e1");
withRoot("e2");
withRoot("e3");
```
but I'm not sure how we'd tie them all to something in the metamodel to be used in that util... 😕

Reply:
> but I'm not sure how we'd tie them all to something in the metamodel to be used in that util... 😕

Essentially, to go beyond simple entity inheritance, we'll need to introduce union types at the object field level.

That's what I was talking about here:

> Union types at the object field level could be a thing too, but that seems... hard. We'd essentially need `SomeNestedObject_` to extend all its union types. That means multiple inheritance, which means `SomeNestedObject_` must be an interface. Possible, since interfaces can have static fields, but needs investigation.

Reply:
ok next iteration 🙈 I've tried to look at the interfaces and whether we can somehow use them to our advantage. This time, the model is:
```java
@MappedSuperclass
public static class MappedSuperclassThing {
	@Id
	Long id;

	@FullTextField
	String a;
}

@Indexed
@Entity
public static class ContainingA extends MappedSuperclassThing {
	@IndexedEmbedded
	EmbeddedThing1 e1;
	@IndexedEmbedded
	EmbeddedThing1 e2;
}

public static class EmbeddedThing1 {
	@FullTextField
	String a;
	@FullTextField
	String b;
	// some other fields
}

public static class EmbeddedThing2 {
	@FullTextField
	String a;
	// some other fields maybe different from EmbeddedThing1
}

@Entity
@Indexed
public static class ContainingB extends MappedSuperclassThing{
	@IndexedEmbedded
	EmbeddedThing3 e3;
}

public static class EmbeddedThing3 {
	@GenericField
	Integer a;
	// some other fields maybe different from EmbeddedThing1/EmbeddedThing2
}
```
There are string/int fields `a` and some embeddables. Ultimately, we'd want to be able to have a util method as:
```java
public static SearchPredicate utilMethodForPredicate(SearchPredicateFactory<? extends Property_String_a_> factory) {
	return factory.match().field( Property_String_a_.a ).matching( "a" ).toPredicate();
}
```
where we want to accept the factories for "scope roots" that have a string property `a` in them. Which can be based on `ContainingA`/`ContainingB`/`EmbeddedThing1`/`EmbeddedThing2`. For the util to work `a` should be a part of the `Property_String_a_`:
```java
interface Property_String_a_ {
	ValueFieldReference1<Property_String_a_, String, String, String> a = ....;
}
```
then we can have the metamodel as:
```java
public static class MappedSuperclassThing_ implements Property_String_a_ {

	public static HibernateOrmRootReferenceScope<MappedSuperclassThing_, MappedSuperclassThing> scope;

	static {
		//a = ValueFieldReference1.of( "a", MappedSuperclassThing_.class, String.class, String.class, String.class );

		scope = RootReferenceScopeImpl.of( MappedSuperclassThing_.class, MappedSuperclassThing.class );
	}
}

public static class ContainingA_ extends MappedSuperclassThing_ {
	public static e1_.Absolute e1;
	public static e2_.Absolute e2;

	public static HibernateOrmRootReferenceScope<ContainingA_, ContainingA> scope;

	static {
		e1 = new e1_.Absolute();
		e2 = new e2_.Absolute();

		scope = RootReferenceScopeImpl.of( ContainingA_.class, ContainingA.class );
	}

	public static class e1_ implements ObjectFieldReference<ContainingA_>, Property_String_a_, Property_String_b_ {

		@Override
		public String absolutePath() {
			return "e1";
		}

		@Override
		public Class<ContainingA_> scopeRootType() {
			return ContainingA_.class;
		}

		static class Absolute extends e1_ {
			public static ValueFieldReference1<ContainingA_, String, String, String> a;
			public static ValueFieldReference1<ContainingA_, String, String, String> b;

			static {
				a = ValueFieldReference1.of( "e1.a", ContainingA_.class, String.class, String.class, String.class );
				b = ValueFieldReference1.of( "e1.b", ContainingA_.class, String.class, String.class, String.class );
			}
		}
	}

	public static class e2_ implements ObjectFieldReference<ContainingA_>, Property_String_a_, Property_String_b_ {

		@Override
		public String absolutePath() {
			return "e2";
		}

		@Override
		public Class<ContainingA_> scopeRootType() {
			return ContainingA_.class;
		}

		static class Absolute extends e2_ {
			public static ValueFieldReference1<ContainingA_, String, String, String> a;
			public static ValueFieldReference1<ContainingA_, String, String, String> b;

			static {
				a = ValueFieldReference1.of( "e2.a", ContainingA_.class, String.class, String.class, String.class );
				b = ValueFieldReference1.of( "e2.b", ContainingA_.class, String.class, String.class, String.class );
			}
		}
	}
}
```
Since we use `? super SR` we can do
```java
SearchScope<ContainingA_, ContainingA> scope = ContainingA_.scope.create( searchSession );

searchSession.search( scope )
		.select( f -> f.field( ContainingA_.a ) )
		.where( f -> utilMethodForPredicate( f ) )
		.fetchHits( 20 )
```
as `ContainingA_` extends `Property_String_a_` it works both for projection and util.
Hence, I'm thinking - we can create an interface-per-filed, where we reuse the interfaces with the same name and capabilities(traits)...